src/Streamly/Streams/StreamK.hs

```diff
@@ -798,6 +868,22 @@ sequence m = go (toStream m)
             yieldk ma r = unStream (toStream $ ma |: go r) st stp sng yld
          in (unStream m1) (rstState st) stp single yieldk
 
+-------------------------------------------------------------------------------
+-- Inserting
+-------------------------------------------------------------------------------
+
+{-# INLINE intersperseM #-}
+intersperseM :: (IsStream t, MonadAsync m) => m a -> t m a -> t m a
+intersperseM a m = fromStream $ prependingStart (toStream m)
+    where
+    prependingStart m1 = Stream $ \st stp sng yld ->
+        let yieldk i x = unStream (return i |: go x) st stp sng yld
+         in unStream m1 st stp sng yieldk
```

Comment:
We need to use `rstState st` when running m1 here. The `SVar` in `st` is to be shared only at this level (i.e. in the continuations we composed) and not with `m1`.